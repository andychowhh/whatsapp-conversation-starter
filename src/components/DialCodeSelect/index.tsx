import React from 'react';
import Select, {
  components,
  ActionMeta,
  MultiValue,
  OptionProps,
  SingleValue as SingleValueType,
  SingleValueProps
} from 'react-select';
import { Box, HStack } from '@chakra-ui/react';
import CountryFlags from 'country-flag-icons/react/3x2';
import { hasFlag } from 'country-flag-icons';

import { FORMATTED_COUNTRY_CODES } from 'constants/';

import { FormattedCountryCode } from 'types';

import { colorStyles } from './DialCodeSelectStyles';

interface DialCodeSelectProp {
  value: FormattedCountryCode;
  onChange: (
    newValue:
      | SingleValueType<FormattedCountryCode>
      | MultiValue<FormattedCountryCode>,
    actionMeta: ActionMeta<FormattedCountryCode>
  ) => void;
}

const { Option, SingleValue } = components;

const IconOption = (props: OptionProps<FormattedCountryCode>) => {
  const { label, value, dialCode } = props.data;
  const CountryFlag = hasFlag(value)
    ? CountryFlags[value as keyof typeof CountryFlags]
    : () => <></>;

  return (
    <Option {...props}>
      <HStack>
        <Box whiteSpace="nowrap">{label}</Box>
        <Box w="20px" h="20px" display="flex" alignItems="center">
          <CountryFlag />
        </Box>
        <Box>{`(+${dialCode})`}</Box>
      </HStack>
    </Option>
  );
};

const CustomSingleValue = (props: SingleValueProps<FormattedCountryCode>) => {
  const { dialCode } = props.data;
  return <SingleValue {...props}>{`+${dialCode}`}</SingleValue>;
};

export function DialCodeSelect({ value, onChange }: DialCodeSelectProp) {
  return (
    <Select
      options={FORMATTED_COUNTRY_CODES}
      value={value}
      isSearchable={false}
      components={{ Option: IconOption, SingleValue: CustomSingleValue }}
      onChange={onChange}
      styles={colorStyles}
      data-testid="dial-code-select"
    />
  );
}
