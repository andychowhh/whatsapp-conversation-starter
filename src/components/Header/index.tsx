import React from "react";
import { Heading, Stack, Text } from "@chakra-ui/react";

function Header() {
  return (
    <Stack align={"center"}>
      <Heading fontSize={"4xl"} textAlign="center">
        Start Whatsapp Conversation
      </Heading>
      <Text fontSize={"lg"} color={"gray.600"} textAlign="center">
        No need to add an extra person in your contacts!
      </Text>
    </Stack>
  );
}

export default Header;
