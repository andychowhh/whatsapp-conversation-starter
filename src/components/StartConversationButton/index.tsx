import React from 'react';
import { Button } from '@chakra-ui/react';
import { isValidPhoneNumber } from 'react-phone-number-input';

import { WHATSAPP_BASE_URL, COUNTRY_CODES } from 'constants/';

interface StartConversationButtonProp {
  countryCode: string;
  phoneNumber: string;
  setInputError: (error: string) => void;
}

export function StartConversationButton({
  countryCode,
  phoneNumber,
  setInputError
}: StartConversationButtonProp) {
  const onStartConversationButtonClick = (): void => {
    const dialCode = COUNTRY_CODES.find(
      (item) => item.code === countryCode
    )?.dial_code;
    setInputError('');
    if (!isValidPhoneNumber(`+${dialCode}${phoneNumber}`)) {
      setInputError('The phone number is invalid!');
      return;
    }

    window.open(`${WHATSAPP_BASE_URL}/${dialCode}${phoneNumber}`, '_blank');
  };
  return (
    <Button
      colorScheme="teal"
      size="md"
      w="100%"
      onClick={onStartConversationButtonClick}
    >
      Start Conversation
    </Button>
  );
}
