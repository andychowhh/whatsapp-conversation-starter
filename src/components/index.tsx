export * from './DialCodeSelect';
export * from './Header';
export * from './StartConversationButton';
