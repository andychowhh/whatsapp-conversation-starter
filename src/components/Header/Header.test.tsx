import React from 'react';
import { render } from '@testing-library/react';

import { Header } from 'components';

describe('Header', () => {
  it('renders correctly by snapshot', () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });
});
