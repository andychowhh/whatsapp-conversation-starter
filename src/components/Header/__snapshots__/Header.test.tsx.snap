// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly by snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-0"
  >
    <h2
      class="chakra-heading css-0"
    >
      Start Whatsapp Conversation
    </h2>
    <p
      class="chakra-text css-0"
    >
      No need to add an extra person in your contacts!
    </p>
  </div>
</div>
`;
